### PR TITLE
TST sklearn: better test infrastructure and more tests

### PR DESCRIPTION
**Pull request by adrinjalali**
https://github.com/huggingface/api-inference-community/pull/79 would need a better way to test improvements, but it'd make that PR quite large. This PR creates that infrastructure.

It includes a script which would create the repositories used in tests on the hub as well. It can easily be extended to other tasks and cases.

cc @osanseviero @BenjaminBossan @Narsil

### CODE REVIEW

**Review by Narsil — APPROVED** [review]
This seems really nice to be able to create easily test repos.

Anything we can do to further help ?

**Comment by adrinjalali** [comment]
> Anything we can do to further help ?

You could merge if there are no red flags :D

**Comment by osanseviero** [comment]
I'll take a look at the PR as well if that's ok :)

**Comment by adrinjalali** [comment]
> I'll take a look at the PR as well if that's ok :)

that'd be really nice!

File: docker_images/sklearn/tests/generators/iris.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    )
+
+
+if __name__ == "__main__":
+    version = sys.argv[1]
+
+    X, y = load_iris(return_X_y=True, as_frame=True)
+    X_train, X_test, y_train, y_test = train_test_split(
+        X, y, test_size=0.2, random_state=42
+    )
```
Comment (osanseviero) [inline docker_images/sklearn/tests/generators/iris.py:56]: Why not directly train with X and y? Or if you don't plan to upload metrics, maybe

```suggestion
    X_train, _, y_train, _ = train_test_split(
        X, y, test_size=0.2, random_state=42
    )
```

File: docker_images/sklearn/tests/generators/iris.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    with open(pkl_name, mode="bw") as f:
+        pickle.dump(est, file=f)
+
+    local_repo = mkdtemp(prefix="skops-")
+    hub_utils.init(
+        model=pkl_name,
+        requirements=[f"scikit-learn={sklearn.__version__}"],
+        dst=local_repo,
+        task="tabular-classification",
+        data=X,
```
Comment (osanseviero) [inline docker_images/sklearn/tests/generators/iris.py:70]: (out of curiosity and totally unrelated to this PR, but what's the use of `data`?)

File: docker_images/sklearn/tests/test_api_tabular_classification.py [hunk 3]
```diff
@@ -38,19 +46,26 @@ def tearDown(self):
...
-        self.assertEqual(
-            response.status_code,
-            200,
-        )
+
+        assert response.status_code == 200
         content = json.loads(response.content)
-        self.assertEqual(type(content), list)
-        self.assertEqual(len(content), expected_output_len)
+        assert type(content) == list
```
Comment (osanseviero) [inline docker_images/sklearn/tests/test_api_tabular_classification.py:67]: Why are you using asserts instead of `self.assertEqual` for consistency with the rest of this repo's codebase?

**Review by osanseviero — APPROVED** [review]
Very nice and useful! Huge benefit of being able to train models in seconds with `sklearn` :D I left some minor comments but this looks great

File: docker_images/sklearn/tests/generators/iris.py (hunk 1 above)
Comment (adrinjalali) [reply docker_images/sklearn/tests/generators/iris.py]: this is used partly to shuffle the data. And the `X_test` is used bellow.

File: docker_images/sklearn/tests/generators/iris.py (hunk 2 above)
Comment (adrinjalali) [reply docker_images/sklearn/tests/generators/iris.py]: two things:
- get the first three rows to save in `config.json` and the README metadata's `widget` section for widgets to work.
- stores the columns and their order in `config.json` so that the bakend can construct a correct data to be passed to the model. See https://github.com/huggingface/api-inference-community/pull/79/files

File: docker_images/sklearn/tests/test_api_tabular_classification.py (hunk 3 above)
Comment (adrinjalali) [reply docker_images/sklearn/tests/test_api_tabular_classification.py:67]: because we use `pytest` and it correctly understands `assert`. I find `assert x == y` significantly more readable and elegant than `self.assertEqual(x, y)`.

"We should do it the XXX way cause we've done it this way" is not really a good argument ;)

File: docker_images/sklearn/tests/generators/iris.py (hunk 1 above)
Comment (osanseviero) [reply docker_images/sklearn/tests/generators/iris.py]: Ah alright! It was passing `X` before hehe

File: docker_images/sklearn/tests/generators/iris.py [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+from sklearn.linear_model import LogisticRegression
+from sklearn.model_selection import train_test_split
+from sklearn.pipeline import make_pipeline
+from sklearn.preprocessing import StandardScaler
+from skops import hub_utils
+
+
+def push_repo(repo_name, local_repo):
+    # this token should be allowed to push to the skops-tests org.
+    token = os.environ["SKOPS_TESTS_TOKEN"]
```
[the remark this answers is not in the document]
Comment (adrinjalali) [reply docker_images/sklearn/tests/generators/iris.py:22]: The readme says: 


> You can make sure those repos are up to date by running the
> `docker_images/sklearn/tests/generators/run.sh` script. The script creates
> required conda environments and runs scripts inside those environments. You
> should also give it a valid token with access to the `skops-tests` org:
> 
> ```
> SKOPS_TESTS_TOKEN=your_secret_token path/to/run.sh
> ```

The user wouldn't run the python file, the script runs it.

File: docker_images/sklearn/tests/test_api_tabular_classification.py (hunk 3 above)
Comment (osanseviero) [reply docker_images/sklearn/tests/test_api_tabular_classification.py:67]: Consistency in a codebase is more important to me, especially as other library owners look into other implementations when implementing theirs. Up to you though!

Comment (adrinjalali) [reply docker_images/sklearn/tests/test_api_tabular_classification.py:67]: They should start using `assert` too :D

Comment (Narsil) [reply docker_images/sklearn/tests/test_api_tabular_classification.py:67]: Actually no, `assert` is not great because we cannot enforce the use `pytest`. `python -m unittest` works just as well.
And if you're not using it, then the error messages are pretty unreadable.

Also `transformers` is also using that format, and pretty much every lib in the HF ecosystem I have seen so far.
It's just the most portable one, so let's go for this one, please.

I won't strongly oppose this as long as the test does its job, but I feel more favorable to `assertEqual`.

File: docker_images/sklearn/tests/generators/iris.py (hunk 2 above)
Comment (osanseviero) [reply docker_images/sklearn/tests/generators/iris.py]: Very cool, thanks for explaining!

File: docker_images/sklearn/tests/test_api_tabular_classification.py (hunk 3 above)
Comment (adrinjalali) [reply docker_images/sklearn/tests/test_api_tabular_classification.py:67]: Note that these tests do use pytest. It's importing it, so w/o `pytest` this PR doesn't work anyway. And we pretty much never ask people to run tests with unittest, we always give them a pytest command to run.

Comment (Narsil) [reply docker_images/sklearn/tests/test_api_tabular_classification.py:67]: If I am not mistaken, you are adding the hard dependency, otherwise the file would be relatively agnostic.

You are free to break consistency here as `sklearn` inference is meant to be isolated from the rest, just commenting about the rest of HF ecosystem.

I'll resolve as I think there's not really anything to add anymore.

**Comment by adrinjalali** [comment]
Shall we merge this?

**Comment by osanseviero** [comment]
Yes, merging!
